127. Word Ladder/WordLadderStep3.java

```diff
@@ -0,0 +1,71 @@
+public class WordLadderStep3 {
+    // 12min
+    class Solution {
+        record WordAndCount(String word, int count) {}
+
+        public int ladderLength(String beginWord, String endWord, List<String> wordList) {
+            List<String> searchWords = new ArrayList<>();
+            searchWords.addAll(wordList);
+            searchWords.add(beginWord);
+            Map<String, List<String>> wordToAdjacentWords = new HashMap<>();
+            for (String word : searchWords) {
+                wordToAdjacentWords.put(word, new ArrayList<>());
+            }
+            for (int i = 0; i < searchWords.size(); i++) {
+                String word1 = searchWords.get(i);
+                for (int j = i + 1; j < searchWords.size(); j++) {
+                    String word2 = searchWords.get(j);
+                    if (isOneLetterDifferent(word1, word2)) {
+                        wordToAdjacentWords.get(word1).add(word2);
+                        wordToAdjacentWords.get(word2).add(word1);
+                    }
+                }
+            }
+
+            Queue<WordAndCount> wordAndCounts = new LinkedList<>();
+            wordAndCounts.offer(new WordAndCount(beginWord, 1));
+            Set<String> usedWords = new HashSet<>();
+            usedWords.add(beginWord);
+            while (!wordAndCounts.isEmpty()) {
+                WordAndCount wordAndCount = wordAndCounts.poll();
+                String word = wordAndCount.word();
+                int count = wordAndCount.count();

```

Comment:
count は常に 1 が加算されるので、宣言時に加算してしまってもいいかもと思いました。
```suggestion
                int count = wordAndCount.count() + 1;
```

Reply:
ありがとうございます。
確かに仰るように必ずインクリメントされていますね。
ここのタイミングでは ladder できる単語がまだ見つかっていないので、個人的に count() のままの方が安心しちゃいます。でも、まとめた方がミスが少なそうです。